dev-tools/scripts/refguide/gen-refguide-redirects.py

```diff
@@ -0,0 +1,134 @@
+    parser.add_argument('--old', required=True, help='Old pagenames file, one .adoc filename per line')
+    parser.add_argument('--new', required=True, help='New pagenames file, one .adoc filename per line')
+    parser.add_argument('--mapping', required=True, help='Semicolon separated from-to file names (adoc)')
+    parser.add_argument('--htaccess', action='store_true', default=False, help='Output as htaccess rules')
+    newconf = parser.parse_args()
+    return newconf
+
+
+def out(text):
+    global conf
+    if not conf.htaccess:
+        print(text)
+
+
+def lines_from_file(filename):
+    with open(filename, 'r') as fp:
+        lines = []
+        for line in fp.readlines():
+            if line.startswith("#") or len(line.strip()) == 0:
+                continue
+            lines.append(line.replace(".adoc", ".html").strip())
+        return lines
+
+
+def main():
+    global conf
+    conf = read_config()
+
+    new = {}
+    name_map = {}
+
+    out("Reading config")
+    old = lines_from_file(conf.old)
+    for line in lines_from_file(conf.new):
+        (path, file) = line.split("/")
+        new[file] = line
+    for line in lines_from_file(conf.mapping):
+        (frm, to) = line.split(";")
+        name_map[frm] = to
+
+    # Files in src/old-pages as of 2022-02-04
+    old_pages = ["configuration-apis.html", "configuration-guide.html", "controlling-results.html", "deployment-guide.html", "enhancing-queries.html", "field-types.html", "fields-and-schema-design.html", "getting-started.html", "indexing-data-operations.html", "installation-deployment.html", "monitoring-solr.html", "query-guide.html", "scaling-solr.html", "schema-indexing-guide.html", "solr-concepts.html", "solr-schema.html", "solrcloud-clusters.html", "user-managed-clusters.html"]
+
+    result = {}
+    old_guide = []
+    failed = {}
+    regex_new = {}
+    out("Converting...")
+    for frm in old:
+        if frm in new:
+            (subpath, name) = new[frm].split("/")
+            if subpath not in regex_new:
+                regex_new[subpath] = []
+            regex_new[subpath].append(name.split(".html")[0])
+        elif frm in name_map:
+            new_name = name_map[frm]
+            if new_name in new:
+                result[frm] = new[new_name]
+            elif new_name.startswith("/guide/"):
+                result[frm] = new_name[7:]
+            elif new_name == "_8_11":
+                old_guide.append(frm.split(".html")[0])
+            else:
+                failed[frm] = "Mapped value %s not in new guide" % new_name
+        elif frm in old_pages:
+            failed[frm] = "Not yet mapped (in src/old-pages)"
+        else:
+            failed[frm] = "404"
+
+    if conf.htaccess:
+        print("# Existing pages moved to sub path")
+        for key in regex_new:
+            print("RedirectMatch ^/guide/(%s)\.html /guide/%s/$1.html" % ("|".join(regex_new[key]), key))
```

Comment:
The new Solr Ref Guide pages are going to have the structure:  `/guide/solr/<version>/page.html`. (This includes a `/solr/` after `/guide/`). And the guide will support the version tag `latest` that will redirect to the latest version that has been published. Therefore these should link to `/guide/solr/latest/<page>.html` (this line and the rewrite rule below it)

Reply:
Ok, will add the `/solr/` level, and `latest` version.

Reply:
Done. I presume that the already existing guides will continue to live at `/guide/<version>/`?

Reply:
Yes, that presumption is correct I think.

Also, Antora is going to create another htaccess file to redirect stuff itself (including `/guide/solr/latest` -> `/guide/solr/9_0`). Everytime I look at htaccess, I've forgotten what I learned the previous time. Will these new rules make it so that the browser does two redirects, `old-url` -> `/guide/solr/latest` and then the browser issues another request: `/guide/solr/latest` -> `/guide/solr/9_0`? Or will htaccess do the 2 redirects itself and give the browser `old-url` -> `/guide/solr/9_0`?

Reply:
This question is not a blocker, at least not for now. Easiest path forward to make this stuff work, I just wanted to better understand.

Reply:
Multi layers of .htaccess is hard to get right. The Antora folders will all reside in svn in the folder pointed to by `__root/docs.solr.apache.org/guide`, so I don't even know if httpd will execute .htaccess from there. If httpd finds a `.htaccess` file on a deep level, it will use that one instead of the topmost one, and if you try to inherit, then all the rules form top-level stop working.

So I think we have to put all rules in our main htaccess file, including the `latest` -> `9_0` one. At the end, we'll simply have to try and see what happens.